packages/lib-bitcoinjs/index.ts

```diff
@@ -23,13 +28,38 @@ const assertSigned = (res: GenerateSignatureResponse) => {
   }
 }
 
+const getTapInternalKey = (publicKey: Buffer, taprootMerkleRoot?: Buffer, parityByte?: boolean) => {
+  const schnorrPublicKey = publicKey.subarray(1, 33)
+  const tweak = taggedHash('TapTweak', Buffer.concat([schnorrPublicKey, taprootMerkleRoot || Buffer.alloc(0)]))
+  const tweakedPublicKey = ecc.xOnlyPointAddTweak(schnorrPublicKey, tweak)
+  if (parityByte) {
+    return Buffer.concat([
+      Uint8Array.from([tweakedPublicKey!.parity === 0 ? 0x02 : 0x03]),
+      tweakedPublicKey!.xOnlyPubkey,
+    ])
+  } else {
+    return Buffer.concat([tweakedPublicKey!.xOnlyPubkey])
+  }
+}
```

Comment:
this function is unnecessary, `bitcoinjs` already has this utility function.

```
import { tweakKey } from 'bitcoinjs-lib/src/payments/bip341'
```

instead, let's add a new getter/property to `DfnsWallet`, then it's easy to tweak the public key `tweakKey(wallet.xOnlyPublicKey, taprootMerkleRoot)`

```
public get xOnlyPublicKey() {
  return this.publicKey.subarray(1)
}
```

Reply:
If I understand your message correctly, you want our customers to derive the public key themselves. I don't think we can take this path:

- Because we are implementing `SignerAsync`, we need the tweaked signer to have a tweaked public key, so we need to compute the new public key anyway
- The library does not allow the hash to be empty: https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/cjs/payments/bip341.cjs#L158, unfortunately, for the default address (empty merkleroot) we cannot use it

Reply:
it [does support](https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/cjs/payments/bip341.cjs#L146) empty/null/undefined merkleroot, here's the test

```
const key = ECPair.fromPrivateKey(
  Buffer.from(
    "bff66a7299bd8bcd1e9ddbe51e5aadf4b51a61da474b55c23dcd1d87b2d04264",
    "hex"
  )
);

console.log(tweakKey(key.publicKey.subarray(1), undefined));

console.log(
  tweakKey(
    key.publicKey.subarray(1),
    Buffer.from(
      "0000000000000000000000000000000000000000000000000000000000000000",
      "hex"
    )
  )
);
```

> If I understand your message correctly, you want our customers to derive the public key themselves.

scratch the idea of `xOnlyPubKey` then, I didn't consider the root wallet also needs to return the tweaked public key, we can change the publicKey to conditionally return either the original one for ecdsa or tweaked one for schnorr, we don't ever need to expose the original if schnorr right?

```
private _publicKey = Buffer.from(publicKey, 'hex)

public get publicKey() {
  if (Schnorr) {
    return tweakKey(this._publicKey.subarray(1), undefined)
  } else {
    return this._publicKey
  }
}
```

Reply:
> If I understand your message correctly, you want our customers to derive the public key themselves.

I guess it's still unclear to me, should `publicKey` be the original key or the tweak key for a Dfns wallet. does the property get used at all when `signInputAsync(signer)` is called?

> Because we are implementing SignerAsync, we need the tweaked signer to have a tweaked public key, so we need to compute the new public key anyway

justification in my other comment, tweaked signer should be completely independent from the `DfnsWallet` class, because I find mixing them into one class is confusing. the behavior of the class becomes hard to reason through. also we can compute the key by using the bitcoin js helper. my original comment is we don't need to duplicate the logic.

> The library does not allow the hash to be empty

it [does support](https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/cjs/payments/bip341.cjs#L146) empty/null/undefined merkleroot, here's the test

```
const key = ECPair.fromPrivateKey(
  Buffer.from(
    "bff66a7299bd8bcd1e9ddbe51e5aadf4b51a61da474b55c23dcd1d87b2d04264",
    "hex"
  )
);

console.log(tweakKey(key.publicKey.subarray(1), undefined));

console.log(
  tweakKey(
    key.publicKey.subarray(1),
    Buffer.from(
      "0000000000000000000000000000000000000000000000000000000000000000",
      "hex"
    )
  )
);
```

Reply:
> I guess it's still unclear to me, should publicKey be the original key or the tweak key for a Dfns wallet

When the signer is tweaked, it needs to return the tweaked public key, `signInputAsync` will call `signer.publicKey` we have no choice but to modify it

>  tweaked signer should be completely independent from the DfnsWallet class

I don't fully agree, maybe we can huddle about it

> it [does support](https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/cjs/payments/bip341.cjs#L146) empty/null/undefined merkleroot, here's the test

my bad, it is now fixed